cmd/zoekt-sourcegraph-indexserver/queue.go

```diff
@@ -36,6 +38,12 @@ type queueItem struct {
 	// dateAddedToQueue is the time when this indexing job was added to the queue. If this item is no longer
 	// in the heap (i.e. it has been processed already), this value is nonsensical.
 	dateAddedToQueue time.Time
+	// consecutiveFailures is the count of preceding consecutive failures. Backoff duration upon indexing failure increases linearly
+	// with this count,
+	consecutiveFailures int
+	// backOffUntil is the earliest time when we allow the item to be pushed to the heap. Until then the item will not be enqueued
+	// and indexing will not be attempted.
+	backoffUntil time.Time
```

Comment:
Initial solution I coded was to use a counter identify subsequent enqueue operations to skip/not fulfill. We'd decrement the counter whenever we skip. We only enqueue again once counter reaches 0. An indexing failure would set a new positive value for the counter; this value would increase linearly with the number of consecutive failures (up to a max duration which `Queue` is aware of).

I believe the counter approach is an even simpler solution since we don't deal with time and we're already obfuscating how long indexing will be paused for/ how long we back off indexing attempts for. However, it's not intuitive since we don't think in terms of enqueue requests. A given sync operation could attempt to enqueue up to two times from what I can tell (because we'd request enqueue once for `IterateIndexOptions()` and then again for `Bump()`, so counter doesn't represent Sync cycles skipped but rather enqueue requests.

Reply:
yeah unfortunetly we can't rely on just counters here, time is a good thing to track instead. We regularly call thingsl ike bump and enqueue.